File: tools/repair_ws_pools.py
import re
import json
from pathlib import Path

repo = Path(__file__).resolve().parents[1]
json_path = repo / 'json' / 'ws-pools.json'
js_path = repo / 'js' / 'ws-pools.js'

def extract_arrays(text):
    pattern = re.compile(r'["\']([A-Za-z0-9_\-]+)["\']\s*:\s*\[([^\]]*)\]')
    result = {}
    for m in pattern.finditer(text):
        key = m.group(1)
        inner = m.group(2)
        items = re.findall(r'["\']([^"\']+)["\']', inner)
        if not items:
            parts = [p.strip().strip('"\'') for p in inner.split(',') if p.strip()]
            items = [p for p in parts if p]
        if items:
            result.setdefault(key, []).extend(items)
    return result

# Build a letter-token regex: prefer the `regex` module for \p{L}, otherwise fall back
try:
    import regex
    letter_re = regex.compile(r"\p{L}+", regex.VERSION1)
except Exception:
    # fallback: approximate with Latin letters and common accented ranges
    letter_re = re.compile(r"[A-Za-zÀ-ÖØ-öø-ÿ]+")


def extract_letter_tokens(arr):
    tokens = []
    for it in arr:
        if not it: continue
        s = str(it).upper()
        parts = letter_re.findall(s) or []
        for p in parts:
            if p and len(p) >= 2:
                tokens.append(p)
    return tokens


def merge_and_clean(maps):
    merged = {}
    for mp in maps:
        for k,v in mp.items():
            merged.setdefault(k, [])
            merged[k].extend(v)
    for k in list(merged.keys()):
        raw = extract_letter_tokens(merged[k])
        seen = set(); cleaned = []
        for t in raw:
            up = t.upper()
            if up not in seen:
                seen.add(up); cleaned.append(up)
        norm = []
        for t in cleaned:
            chosen = t
            shorter = [s for s in cleaned if s != t and len(s) >= 3 and t.startswith(s)]
            if shorter:
                shorter.sort(key=lambda x: -len(x)); chosen = shorter[0]
            else:
                for drop in range(1, min(5, len(t))):
                    cand = t[:-drop]
                    if len(cand) >= 3 and cand in cleaned:
                        chosen = cand; break
            if chosen not in norm:
                norm.append(chosen)
        merged[k] = [x for x in norm if 3 <= len(x) <= 20]
    return merged


def write_json_pools(pools, path):
    obj = { 'pools': pools }
    path.write_text(json.dumps(obj, indent=2, ensure_ascii=False) + '\n', encoding='utf-8')


def write_js_pools(pools, path):
    data = { 'pools': { k: [s.lower() for s in v] for k,v in pools.items() } }
    content = f"// Embedded WordSearch pools (repaired)\nwindow.WS_POOLS = {json.dumps(data, indent=2, ensure_ascii=False)};\n"
    path.write_text(content, encoding='utf-8')


def main():
    maps = []
    for p in [json_path, js_path]:
        try:
            txt = p.read_text(encoding='utf-8')
            mp = extract_arrays(txt)
            maps.append(mp)
            print('Extracted from', p.name, list(mp.keys()))
        except Exception as e:
            print('Error reading', p, e)
    pools = merge_and_clean(maps)
    print('Final categories:', list(pools.keys()))
    for k in pools:
        print(k, ':', len(pools[k]), 'tokens (sample):', pools[k][:8])
    try:
        write_json_pools(pools, json_path)
        print('Wrote cleaned JSON to', json_path)
    except Exception as e:
        print('Failed write JSON', e)
    try:
        write_js_pools(pools, js_path)
        print('Wrote cleaned JS to', js_path)
    except Exception as e:
        print('Failed write JS', e)

if __name__ == '__main__':
    main()

File: tools/repair_ws_pools.js
const fs = require('fs');
const path = require('path');

function extractArraysFromText(text){
  const map = {};
  // match "key": [ ... ] blocks (non-greedy for bracket content)
  const re = /["']([A-Za-z0-9_\-]+)["']\s*:\s*\[([^\]]*)\]/g;
  let m;
  while((m = re.exec(text))){
    const key = m[1];
    const inner = m[2];
    const items = [];
    // match strings inside quotes
    const sq = /["']([^"']+)["']/g;
    let s;
    while((s = sq.exec(inner))){
      items.push(s[1]);
    }
    // if no quoted strings found, try comma-separated tokens
    if(items.length === 0){
      const parts = inner.split(',').map(x=>x.trim()).filter(Boolean);
      parts.forEach(p=>{
        // remove trailing commas/quotes
        const cleaned = p.replace(/^["'`\s]+|["'`\s]+$/g,'');
        if(cleaned) items.push(cleaned);
      });
    }
    if(items.length) map[key] = (map[key] || []).concat(items);
  }
  return map;
}

function extractLetterTokens(arr){
  const tokens = [];
  for(const it of arr){
    if(!it) continue;
    const s = String(it).toUpperCase();
    const parts = s.match(/\p{L}+/gu) || [];
    for(const p of parts){ if(p && p.length>=2) tokens.push(p); }
  }
  return tokens;
}

function mergeAndClean(maps){
  const merged = {};
  for(const mp of maps){
    for(const k of Object.keys(mp)){
      merged[k] = merged[k] || [];
      merged[k] = merged[k].concat(mp[k]);
    }
  }
  // clean each category
  for(const k of Object.keys(merged)){
    const raw = extractLetterTokens(merged[k]);
    // dedupe preserving order
    const seen = new Set();
    const cleaned = [];
    for(const t of raw){
      const up = t.toUpperCase();
      if(!seen.has(up)) { seen.add(up); cleaned.push(up); }
    }
    // try to reduce tokens with trailing garbage: if token starts with a shorter token, prefer shorter
    const norm = [];
    for(const t of cleaned){
      let chosen = t;
      const shorter = cleaned.filter(s => s !== t && s.length >= 3 && t.startsWith(s));
      if(shorter.length){ shorter.sort((a,b)=> b.length - a.length); chosen = shorter[0]; }
      else { for(let drop=1; drop<=4 && drop < t.length; drop++){ const cand = t.slice(0, t.length-drop); if(cand.length>=3 && cleaned.includes(cand)){ chosen = cand; break; } } }
      if(!norm.includes(chosen)) norm.push(chosen);
    }
    // final filter: reasonable lengths (3..20)
    merged[k] = norm.filter(x=> x.length >= 3 && x.length <= 20);
  }
  return merged;
}

function writeJsonPools(pools, filePath){
  const obj = { pools };
  fs.writeFileSync(filePath, JSON.stringify(obj, null, 2) + '\n', 'utf8');
}

function writeJsPools(pools, filePath){
  const data = {
    pools: {}
  };
  for(const k of Object.keys(pools)){
    // write lowercase tokens for js file readability
    data.pools[k] = pools[k].map(s=> s.toLowerCase());
  }
  const content = `// Embedded WordSearch pools (repaired)\nwindow.WS_POOLS = ${JSON.stringify(data, null, 2)};\n`;
  fs.writeFileSync(filePath, content, 'utf8');
}

function main(){
  const repo = path.resolve(__dirname, '..');
  const jsonPath = path.join(repo, 'json', 'ws-pools.json');
  const jsPath = path.join(repo, 'js', 'ws-pools.js');
  const files = [jsonPath, jsPath];
  const maps = [];
  for(const f of files){
    try{
      const txt = fs.readFileSync(f, 'utf8');
      const mp = extractArraysFromText(txt);
      maps.push(mp);
      console.log('Extracted from', path.basename(f), Object.keys(mp).join(','));
    }catch(e){ console.error('Error reading', f, e.message); }
  }
  const pools = mergeAndClean(maps);
  console.log('Final categories:', Object.keys(pools));
  // show counts
  Object.keys(pools).forEach(k=> console.log(k, ':', pools[k].length, 'tokens (sample)', pools[k].slice(0,8).join(', ')));

  // write back
  try{ writeJsonPools(pools, jsonPath); console.log('Wrote cleaned JSON to', jsonPath); }catch(e){ console.error('Failed write JSON', e.message); }
  try{ writeJsPools(pools, jsPath); console.log('Wrote cleaned JS to', jsPath); }catch(e){ console.error('Failed write JS', e.message); }
}

main();
